js/python-main.js

```diff
@@ -603,6 +623,33 @@ function web_editor(config) {
       });
     }
 
+    function webkitHostMsgFile(name,data,action) {
+        // 'host' refers to the app embedding the editor. We use 'host' here to 
+        // match the MakeCode messageHandler name
+        window.webkit.messageHandlers.host.postMessage({'name': name,action:data});
```

Comment:
@martinwork The other messages sent from the editor to classroom all have a mandatory `type` key in the interface.

Is this message interface what MakeCode already sends to the app? I've struggled to find a postMessage in the pxt repo with that signature.

Reply:
Also I don't think the argument `action` is used here?
Is it meant to set the object key for `data`?

This:
```js
function foo(name, data, action) {
     console.log({'name': name, action:data});
}
foo('one', 'two', 'three')
```
Prints: `{name: "one", action: "two"}`

Is it meant to do `{name: "one", three: "two"}`?

Reply:
Sorry, the code was completely untested, and written by a javascript beginner! I guess "action" was a bad name choice.

I offered the code to ask if adding messages like this would be acceptable, and said if so, I would test it. 

GitHub search isn't always very helpful is it? I think the pxt code has changed since I last looked.
https://github.com/microsoft/pxt/blob/master/pxteditor/editor.ts#L554
https://github.com/microsoft/pxt/blob/master/webapp/src/cmds.ts#L129
https://github.com/microsoft/pxt/blob/master/webapp/src/cmds.ts#L146
https://github.com/microsoft/pxt/blob/master/webapp/src/cmds.ts#L158

I'm not attempting to build a single view to host either MakeCode or Python, so feel free to modify the format of the messages. I think the message handler just extracts the parameters and passes them to another function anyway.

{ name: "my.hex", [download|save]: "whatever" } is a little awkward to handle.

{ type: "upysomething" reason: "download", name: "my.hex", contents:  "whatever" } and
{ type: "upysomething" reason: "save", name: "my.hex", contents:  "whatever" } might be simpler.

Reply:
Thanks Martin, those pxt links were great to understand the differences!
I agree, I think it might be better to try to continue using the message formatting that we have with classroom as you've listed.

@microbit-sam I assume this is fine as well for the Android app? Does it currently use any postMessage with MakeCode?